### DISCUSSION TITLE
FROZEN_KV_MTP slower than no-spec on Gemma 4 (SWA hybrid) under concurrency

### DISCUSSION THREAD

**Discussion by midakuyo** [sgl-project/sglang#36314] [General]
**Environment**
- sm_80 (CMP 170HX, GA100 64GB) · triton attention backend (forced for Gemma 4)
- Model: [lokeshe09/gemma-4-31B-it-INT8](https://huggingface.co/lokeshe09/gemma-4-31B-it-INT8) (compressed-tensors W8A8)
- Draft: [google/gemma-4-31B-it-assistant](https://huggingface.co/google/gemma-4-31B-it-assistant) (official, BF16)
- `--page-size 64` · lmsysorg/sglang:latest (v0.5.18)

**Symptom**

MTP (NEXTN → FROZEN_KV_MTP) is roughly on par with no-spec for a single request,
but drops to less than half of no-spec under concurrency. (For reference, vLLM's
MTP with the same model, draft, and batch 8 gave +77%.)

| config | solo tg | batch-8 warm tg (per-request median) |
|---|---|---|
| no-spec | 34.3 | **25.5** |
| MTP s1/d2 | — | 9.4 |
| MTP s2/d3 | 34.9 | 11.5 |
| MTP s3/d4 | — | 12.2 |

(~2.4–3.3k-token contexts, ~160-token generations, sustained load with responses
appended to history each round. Full raw logs in the gist below.)

## Observation 1 — `draft` dominates `scheduler.run_batch`

Captured a 6s torch profile (`/start_profile`) mid-load at batch 8 and aggregated
`user_annotation` spans (34–36 iterations sampled):

| steps | run_batch mean | draft mean (share) | TARGET_VERIFY |
|---|---|---|---|
| 1 | 173.4ms | 165.8ms (96%) | 0.5ms |
| 2 | 164.4ms | 155.9ms (95%) | 0.4ms |
| 3 | 153.8ms | 144.5ms (94%) | 0.3ms |

GPU utilization stayed at 97–99% against the 175W power cap during load
(1s sampling) — this looks like real GPU work, not a host-side stall. Verify is
under 1ms, essentially free.

## Observation 2 — iteration time is depth-independent; tg tracks accept length

| steps/draft | accept len (server logs) | batch-8 tg | iteration time (profiler, direct) |
|---|---|---|---|
| 1/2 | 1.65–1.68 | 9.4 | 173ms |
| 2/3 | 1.95–2.15 | 11.5 | 164ms |
| 3/4 | 2.23–2.52 | 12.2 | 154ms |

Acceptance rates are in a normal range per depth (0.42–0.68), so this doesn't
look like a draft-quality problem. Tripling the depth barely changes iteration
time, which suggests a **fixed once-per-iteration cost inside the draft phase**
rather than the draft loop itself (tg / accept-len = iteration rate is constant
at 5.2–5.6 iter/s).

## Observation 3 — the cost scales with batch × min(context, SWA window)

Batch 8 fixed, context sweep (warm tg medians):

| context (tokens) | ~220 | ~670 | ~1270 | ~1870 | ~2460 |
|---|---|---|---|---|---|
| tg | 21.5 | 14.2 | 13.4 | 12.9 | 12.2 |

It degrades steeply up to around the window size (1024) and is nearly flat
beyond it. Batch 4 at 2k→4k is also near-flat (21→18.5), so it doesn't look
like a read of the full context. With context fixed at 2048, batch 1/2/4/8
gives 34 / 29.5 / 20.5 / 12.3 — degradation proportional to batch size.

Taken together, this looks like a cost that **reassembles the window-sized
sliding-window KV/index state on every iteration**. Since the window only
slides by the number of accepted tokens per iteration, an incremental update
seems possible. I did check the two related flags:
`--speculative-attention-mode decode` made no difference, and
`--speculative-draft-window-size` is documented as ignored on this path.

## Question

Is this a known cost? (#29858 seems to be reworking the same buffers from the
correctness side, and the #23705 adaptive-spec roadmap looks like a mitigation
for it.)

---

This is my first question here. To be honest, GitHub is a bit of an intimidating
place for me, so I wanted to ask in Discussions before filing an issue. I did
look through the Gemma 4 related roadmap, but I may well have misunderstood
parts of it (this is my first time building an AI serving setup).

I had AI assistance during the investigation and testing. That said, every
number comes from saved server logs and profiler traces, and the repro scripts
plus raw logs are in this gist:
https://gist.github.com/midakuyo/4dbe0cbb009d3718b8f1a41f150a5131

I hope this doesn't read as slop. Thank you for reading.

**Comment by midakuyo** [comment]
#35592 #36372 
It seems they may be handling a related issue

**Comment by KushagraKanaujia** [comment]
@midakuyo Heads-up: #35592 and #36372 were closed without merging. v0.5.19 shipped #34462, which bounds the Triton sliding-window extend-attention KV loop. I can't confirm it touches FROZEN_KV_MTP, but rerunning your batch-8 grid on v0.5.20 is a cheap check.

I built Throttle for exactly this comparison: it says whether a config is actually cheaper or within noise.

```sh
throttle check --url http://localhost:30000/v1 --gpu-hourly-rate <rate> --prompts chats.jsonl --concurrency 8 --requests-per-block 8 --max-tokens 160 --label mtp-s3d4
```

https://github.com/KushagraKanaujia/throttle

Happy to set it up with you, free.

**Comment by flymysql** [comment]
## Your hypothesis is right about *where* the cost is, and the mechanism is visible in the kernel: the window rebuild is a per-iteration, batch-scaled gather — but it is not an "incremental update" the kernel is currently capable of expressing

This is a well-instrumented report, and the three observations that do the work are: iteration time is depth-independent, cost scales with `batch × min(context, window)`, and `draft` dominates `run_batch` while verify is free. I read the current `main` along with your numbers. Your conclusion — a fixed per-iteration cost that reassembles window-sized state — is the correct reading. Here is the mechanism, plus one correction about the flag you tested.

### 1. Why your model ends up on FROZEN_KV_MTP (worth stating, since it explains the fixed cost)

With `google/gemma-4-31B-it-assistant` as the draft, sglang **promotes your algorithm automatically**:

```python
# python/sglang/srt/arg_groups/speculative_hook.py
is_gemma4_draft = any(arch in ("Gemma4AssistantForCausalLM",
                               "Gemma4UnifiedAssistantForCausalLM")
                      for arch in draft_archs)
...
if speculative_algorithm == "NEXTN" or speculative_algorithm == "EAGLE":
    if is_gemma4_draft:
        logger.info("Detected Gemma4AssistantForCausalLM draft; promoting "
                    f"--speculative-algorithm {speculative_algorithm} to FROZEN_KV_MTP.")
        return "FROZEN_KV_MTP"
    return "EAGLE"
```

Note also `EAGLE3` **raises** for this draft arch. So "NEXTN → FROZEN_KV_MTP" isn't something you misconfigured; it's the only supported route, and it routes you onto a draft path whose attention metadata is rebuilt against the *frozen target pool*.

### 2. The fixed per-iteration cost: window metadata is rebuilt from scratch every draft step

The draft loop calls the attention backend's metadata init **inside the step loop**:

```python
# python/sglang/srt/speculative/frozen_kv_mtp_worker_v2.py
for i in range(self.speculative_num_steps):
    ...
    with (self._target_kv_pool_view(forward_batch),
          forward_context(ForwardContext(attn_backend=self.draft_attn_backend))):
        logits_output = self.draft_model_runner.forward(forward_batch).logits_output
```

and for SWA models that metadata init runs the window gather:

```python
# python/sglang/srt/layers/attention/triton_backend.py
if self.sliding_window_size is not None and self.sliding_window_size > 0:
    window_kv_indptr, _, window_kv_lens, _ = update_sliding_window_buffer(
        self.window_kv_indptr, self.kv_index_translator, req_pool_indices,
        self.sliding_window_size, seq_lens, bs,
        token_to_kv_pool=self.token_to_kv_pool,
        window_kv_indices=self.cuda_graph_window_kv_indices)
```

and the buffer that fills it is sized **`max_num_tokens × sliding_window_size`**:

```python
self.cuda_graph_window_kv_indices = torch.zeros(
    (max_num_tokens * self.sliding_window_size), dtype=torch.int64, device=self.device)
```

That is exactly your Observation 3. The work per iteration is proportional to `batch × window`, **not** to how far the window slid. It is a rebuild, not an extension — which is why tripling the draft depth barely moves iteration time (Observation 2): the cost is once per iteration, independent of depth, so `tg ≈ accept_len / iteration_rate` stays flat. Your profiler reading (`draft` 94–96%, verify ~0.4 ms) is consistent with the cost living in the draft's per-step metadata + forward, not the verify pass.

So: **the sliding-window KV/index state is reassembled per iteration, and it is a gather over the window rather than an incremental append.** Your inference was correct.

### 3. One correction: `--speculative-draft-window-size` is not merely "documented as ignored" — it is *structurally* refused on this path, and for a good reason

You said the flag is "documented as ignored on this path." It's stronger than that. The resolution function returns full attention whenever the draft model already has a native window, and warns:

```python
# python/sglang/srt/speculative/spec_utils.py
def resolve_draft_decode_window(model_runner) -> Tuple[int, int]:
    """... Returns (0, 0) -- full draft attention ... also when the draft model
    already has a sliding window of its own: the index builder emits one KV list
    shared by every draft layer, so it cannot express a per-layer window, and the
    draft model's own window is authoritative ..."""
    window_size = int(cfg.speculative_draft_window_size or 0)
    if window_size <= 0:
        return 0, 0
    native_window = getattr(model_runner, "sliding_window_size", None)
    if native_window is not None and native_window > 0:
        if native_window != window_size:
            logger.warning(
                "Ignoring --speculative-draft-window-size=%d: this draft model has a "
                "sliding window of %d, which the attention backend applies per layer. "
                "Draft-decode windowing stays off.", window_size, native_window)
        return 0, 0
    return window_size, int(cfg.speculative_draft_sink_size or 0)
```

Gemma 4's assistant draft reports `get_attention_sliding_window_size() = self.config.sliding_window - 1`, so `native_window > 0` and you fall into the `return 0, 0` branch regardless of the flag. The comment explains why a fix isn't a one-liner: **the index builder emits a single shared KV list for every draft layer, so it cannot express a per-layer window** — and a native per-layer window is authoritative. That's the architectural constraint behind your observation, and it's worth quoting in any feature request.

### 4. Why "just make it incremental" is harder than it looks

Your suggestion (slide by accepted tokens instead of rebuilding the window) is the right direction, but two facts from the code explain why it isn't a small patch:

- **Frozen-KV draft never writes KV.** In the draft path, `forward_batch.out_cache_loc = None` — with the comment *"Frozen draft never writes KV; None signals fill_from to skip the slot."* The draft attends into the **target's** frozen pool via `target_kv_pool_view`. So there's no draft-side cache to append to; the window is a *view* over target KV, and the view's index list is what's being rebuilt.
- **The window gather is a `min(seq_len, window)` computation**, i.e. `window_kv_lens = torch.minimum(seq_lens, sliding_window_size)`, then a cumsum/indptr fill and a gather into `window_kv_indices`. That is a data movement over `batch × window` per iteration — the shape your sweep measured — and it is not something a moving append can reuse unless the *window index list itself* is kept per-request and updated in place, which collides with the shared-per-layer-list constraint above.

### 5. On the related threads you cited

- **#29858** (buffers rework from the correctness side) and the **#23705** adaptive-spec roadmap: consistent with your reading that this is a known-shaped cost being worked on elsewhere. Neither of the comments on your thread changes the mechanism above.
- The comment pointing at **#35592 / #36372** being closed unmerged and **v0.5.19 shipping #34462** ("bounds the Triton sliding-window extend-attention KV loop") is worth acting on, but note what that patch does and does not touch: the **extend**-attention loop bound is on the *target* prefill/extend path, whereas your cost is the *draft* per-step window gather in `update_sliding_window_buffer` feeding the draft forward. A rerun on v0.5.20 is cheap and I'd do it — but if `draft` still dominates `run_batch` with depth-independent iteration time, #34462 has not addressed your mechanism.

### Summary and what I'd measure next

Your diagnosis stands: the cost is a per-iteration, `batch × min(context, window)` reassembly of window state inside the draft path, and the flag that would let you bound it is deliberately neutralized for drafts that already have a native window (because the shared per-layer index list can't express one). Frozen-KV MTP makes this structural, since the draft has no KV of its own to append to.

The one measurement that would most strengthen a feature request: **profile inside `update_sliding_window_buffer`** (the indptr cumsum, and the gather into `window_kv_indices`) separately from the draft model forward. If the gather dominates, the argument is "the window index build is O(batch × window) per iteration and could be made incremental"; if the forward dominates, the window rebuild is a symptom and the real ask is different. That split would make this much easier for a maintainer to triage than the aggregate `draft` span.

*Verified against `main` on 2026-09-25: `speculative_hook.py` (`_resolve_speculative_algorithm_alias`, FROZEN_KV_MTP promotion, EAGLE3 rejection), `speculative/frozen_kv_mtp_worker_v2.py` (`draft`, per-step `_target_kv_pool_view`, `out_cache_loc = None`), `layers/attention/triton_backend.py` (`update_sliding_window_buffer` call sites, `cuda_graph_window_kv_indices` sizing), `kernels/ops/attention/extend_attention.py` (`SLIDING_WINDOW_SIZE` masking and per-tile skip), `speculative/spec_utils.py` (`resolve_draft_decode_window`), `models/gemma4_mtp.py` (`get_attention_sliding_window_size`).*
